Actions (#872)

lexor opened the pull request:
- [x] Artifacts
- [x] Secrets
- [x] Self-hosted runners
- [x] Workflows
- [x] Workflow jobs
- [x] Workflow runs

Documentation: https://developer.github.com/v3/actions/

acrobat commented:
Thanks for working on this @lexor! Let me know when it's ready for review 🚢

fodinabor commented:
What's missing from the developer.github.com documentation is e.g. dispatching:
https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-workflow-dispatch-event

lexor commented:
> What's missing from the developer.github.com documentation is e.g. dispatching:
> [docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-workflow-dispatch-event](https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-workflow-dispatch-event)

Thanks for `dispatches` function but this pull request based on [this documentation](https://developer.github.com/v3/actions/). After this pull request I can work on [this documentation](docs.github.com/en/free-pro-team@latest/rest/reference/actions).

I did not add organization endpoints because marked on documentation. Organization endpoints currently on public beta and can be change.

> Another note: `workflowId` is in most cases allowed to contain the filename of the workflow, so string should be preferred over int

Can you share example cases? I can't find example on docs.

lexor commented:
Hey @acrobat. This pull request awaiting review.

fodinabor commented:
For [Get a workflow](https://developer.github.com/v3/actions/workflows/#get-a-workflow) for example the docs write:

> Gets a specific workflow. You can also replace :workflow_id with :workflow_file_name. For example, you could use main.yml.

So, yeah I guess a string parameter named just workflow would make sense here.

acrobat commented:
@fodinabor the repo dispatch api endpoint is already added in #931

@lexor A general remark before I do a full review, can you update the doc links to directly link to the new documentation? Example: https://docs.github.com/en/free-pro-team@latest/rest/reference/actions

lexor commented:
> @lexor A general remark before I do a full review, can you update the doc links to directly link to the new documentation? Example: [docs.github.com/en/free-pro-team@latest/rest/reference/actions](https://docs.github.com/en/free-pro-team@latest/rest/reference/actions)

Done, I updated.

acrobat reviewed (changes requested):
This comment is also applicable to all occurences of non-query parameters

> Can you replace the non url parameters with a `array $params/parameters `, so our code isn't fixed on a certain set of parameters of github decides to add new options.

Can you also add a docs entry for the added endpoints? The rest of PR looks good 👌

lexor commented on lib/Github/Api/Repository/Actions/Artifacts.php:
```diff
@@ -0,0 +1,83 @@
+     */
+    public function download(string $username, string $repository, int $artifactId, string $format = 'zip')
+    {
+        // TODO(): Get `Location` header for download link.
```
Actually GitHub's download endpoints didn't return any json. Just needed take `Location` header value for download link. I tried already `$this->client->getLastResponse()->getHeader('Location')` but response didn't have `Location` header.

acrobat replied:
Can you do this call with curl and see if the header is present? If not it's a github issue

lexor replied:
If I use "curl" I can see the "location" header.

> λ ~/ curl -i -u lexor:$token https://api.github.com/repos/$username/$repo/actions/runs/$runId/logs
> HTTP/1.1 302 Found
> Status: 302 Found
> Location: https://pipelines.actions.githubusercontent.com/...

acrobat replied:
Can you check in your "test" setup if the 302 redirect is just followed by the client? I think guzzle will by default follow those redirect to the location specified and by that the request/response with the Location header is not returned for the last response.

lexor replied:
Yes, you are right. Guzzle is automatically redirects, `file_put_contents` solved issue.

lexor commented:
> Can you replace the non url parameters with a `array $params/parameters `, so our code isn't fixed on a certain set of parameters of github decides to add new options.

Sure, replaced. Now non url parameters moved on `array $parameters`.

> Can you also add a docs entry for the added endpoints? The rest of PR looks good 👌

I updated functions and test cases. Also added docs for added endpoints.

acrobat commented:
Thanks @lexor! And congrats on your first contribution! 🎉

dereuromark commented:
What I am missing is the enable of a specific workflow
See https://docs.github.com/en/rest/reference/actions#enable-a-workflow

lexor commented:
> What I am missing is the enable of a specific workflow
> See [docs.github.com/en/rest/reference/actions#enable-a-workflow](https://docs.github.com/en/rest/reference/actions#enable-a-workflow)

Endpoint was just added or I missed it. I can create pull request for enable/disable endpoints.

dereuromark commented:
Yeah, might be new.
But I also struggle finding the right syntax.
The docs say this should work:
```php
$url = 'repos/' . rawurlencode($organization) . '/' . rawurlencode($repository) . '/actions/workflows/' . (string)$id . '/enable';

$this->post($url);
```

But it returns Not Found
I made sure to use a token that has permissions for "Workflows".
